### ISSUE TITLE
Add hyperparameter tuning support for fingerprint/feature generation

### ISSUE THREAD

**Issue by thomasATbayer** [Bayer-Group/MotherML#88]
---
title: "Add hyperparameter tuning support for fingerprint/feature generation"
labels: enhancement
---

## Summary

MotherML has solid Optuna-based hyperparameter tuning for models (`get_hyperparameter_space` on every `m_*.py` model wrapper) and for feature selection (e.g. the `threshold` parameter on `MotherSelectFromModel`), but there is no equivalent tuning support for fingerprint/feature generation. Fingerprint parameters (radius, fpSize, useBondTypes, etc.) currently have to be picked manually and are not part of the same `PipelineWithHyperparameterRooting` search space as the rest of the workflow.

## What already exists

**Models** implement `get_hyperparameter_space(X, y, trial, prefix="")`, e.g.:

- `src/mother/ml/models/m_catboost.py`: tunes `bootstrap_type`, `learning_rate`, `random_strength`, `grow_policy`, `max_depth`/`max_leaves`, `loss_function`, etc.
- `src/mother/ml/models/m_randomForest.py`, `m_lasso.py`, `m_tabpfn.py`, `m_tabicl.py`: each define their own Optuna search spaces the same way.

**Feature selection** also participates, e.g. `src/mother/ml/estimators.py`:

```python
def get_hyperparameter_space(self, X, y, trial: Trial, prefix: str = "") -> dict:
    ...
    suggested_params = {prefix + "threshold": trial.suggest_float(prefix + "threshold", 0, 1, step=0.25)}
```

These all plug into `PipelineWithHyperparameterRooting` / `ColumnTransformerWithHyperparameterRooting` / `FeatureUnionWithHyperparameterRooting` (`src/mother/ml/core.py`), which collect nested `step__parameter` search spaces automatically across a pipeline.

## What's missing

`src/mother/feature_generation/core.py` defines fingerprint transformers (`MaccsFingerprints`, `FingerprintsGeneric`, `MorganFingerprints`, and others built via `FingerprintFactory`) as plain `BaseEstimator`/`TransformerMixin` classes with no `get_hyperparameter_space` method and no inheritance from `AbstractMotherPipeline`. This means:

- Fingerprint parameters (e.g. `radius`, `fpSize`, `includeChirality`, `useBondTypes`, `countSimulation`, `includeRingMembership`, ...) are not discoverable by the hyperparameter-rooting mechanism.
- When a fingerprint generation step is included in a `PipelineWithHyperparameterRooting`, Optuna tuning skips it entirely — only the feature-selection and model steps get tuned.
- Users who want to tune, say, Morgan fingerprint `radius`/`fpSize` alongside the model's hyperparameters have to do so manually outside the existing tuning workflow.

## Performance concern: fingerprint tuning is expensive without caching

Unlike model or feature-selection hyperparameters, changing a fingerprint parameter (e.g. `radius`, `fpSize`) means every trial has to recompute the fingerprint for the entire dataset from raw molecules, which is far more expensive than refitting a model or a feature selector. Adding fingerprint parameters to the Optuna search space naively will noticeably slow down tuning, since `PipelineWithHyperparameterRooting` refits the whole pipeline (including fingerprint generation) on every trial even when most trials only change downstream (model/feature-selection) parameters.

To keep this practical:

- **Use `sklearn.pipeline.Pipeline`'s built-in step caching** (the `memory` parameter, e.g. `Pipeline(steps=..., memory=joblib.Memory(location=...))`), which `PipelineWithHyperparameterRooting` already inherits from since it subclasses `skl_pipe.Pipeline`. This caches each step's fitted transformer output keyed on its parameters/input, so repeated trials with unchanged fingerprint parameters reuse the cached fingerprint output instead of recomputing it. This should be surfaced/documented as part of this change, since it's the natural mitigation and requires no new MotherML infrastructure.
- **Use stepwise/quantized search ranges instead of continuous ones** for parameters like `fpSize` (number of bits), e.g. `trial.suggest_int(prefix + "fpSize", 512, 4096, step=512)` or a categorical choice among powers of two (`512, 1024, 2048, 4096`), rather than suggesting every integer in a wide range. This both reflects that fingerprint bit-widths are conventionally chosen from a small set of sizes, and reduces the number of distinct cache entries `Pipeline(memory=...)` needs to keep, improving the cache hit rate across trials.

## Proposed changes

- Implement `get_hyperparameter_space(X, y, trial, prefix="")` (and `default_parameters(prefix)`, mirroring the pattern used by model/feature-selection wrappers) for the fingerprint transformer classes in `src/mother/feature_generation/core.py`, at least for `MorganFingerprints` to start (the most commonly tuned fingerprint), with reasonable default search ranges for `radius`, `fpSize`, and the boolean flags.
- Use stepped/categorical suggestions (not fine-grained continuous ranges) for parameters like `fpSize`, to limit the number of distinct fingerprint configurations trials can produce.
- Consider whether this should live directly on the transformer classes or via a shared mixin (similar to `_CatboostHyperParams`) so it's easy to extend to other fingerprint types (`AtomPairFingerprints`, `RDKitFingerprints`, etc.) later.
- Ensure the new search space integrates cleanly with `PipelineWithHyperparameterRooting`'s existing `step__parameter` prefixing convention, so fingerprint tuning composes with feature-selection and model tuning in one Optuna study.
- Wire up / document `Pipeline(memory=...)` caching (sklearn's existing mechanism) for pipelines that include a tunable fingerprint step, so repeated trials don't recompute fingerprints unnecessarily.
- Document the new capability, the caching recommendation, and the performance tradeoff in `docs/feature_generation.md` alongside the existing fingerprint documentation, and/or `docs/mother/configurator.md` where tuning is discussed.

## Acceptance criteria

- [ ] At least `MorganFingerprints` (and ideally the shared `FingerprintsGeneric` base) exposes a `get_hyperparameter_space` compatible with `PipelineWithHyperparameterRooting`.
- [ ] Fingerprint parameters like `fpSize` are suggested using stepped/categorical values rather than a dense continuous range.
- [ ] Tuning a full pipeline (fingerprint generation + feature selection + model) via Optuna picks up fingerprint parameters using the same `step__parameter` naming convention as other steps.
- [ ] Docs recommend `Pipeline(memory=...)` caching when tuning fingerprint parameters, with an example, and call out the added tuning cost without it.
- [ ] Unit tests cover the new `get_hyperparameter_space` implementation(s), similar to existing model/feature-selection hyperparameter-space tests.
